### PR TITLE
Alien4Cloud go client

### PR DESCRIPTION
**Pull request by laurentganne**
Added a go client for Alien4Cloud REST API.
Added example showing its usage to :
* create and deploy an application from a template in Alien4Cloud catalog
* run a workflow
* undeploy and delete the application

## How to verify

Following READMEs provided under:
* examples/create-deploy-app/
* examples/run-workflow/
* examples/undeploy-delete-app/

first upload the Ystia Forge welcome application in the Alien4Cloud catalog,
then build the first example:

```bash
cd examples/create-deploy-app
go build -o deploy
```

Run this example:
```bash
./deploy -url https://1.2.3.4:8088 \
         -user myuser \
         -password mypasswd \
         -app myapp \
         -template org.ystia.samples.topologies.welcome_basic
```

(here the first suited location is used, you can specify a location using argument `-location` if you prefer).

Check you see logs of the on-going deployment.
Once the deployement is done, check the URL of the deployed web app is printed, and the example exits without error.
Check you can access this Web app URL.

Build the workflow example:

```bash
cd examples/run-workflow
go build -o run
```

Run a workflow stopping the web server :

```bash
./run -url https://1.2.3.4:8088 \
      -user myuser \
      -password mypasswd \
      -app myapp \
      -worflow stopWebServer
```

Check you see logs of the on-going workflow. Once done, check you can't access the Web App URL.

Run a workflow starting the web server :

```bash
./run -url https://1.2.3.4:8088 \
      -user myuser \
      -password mypasswd \
      -app myapp \
      -worflow startWebServer
```

 Check you see logs of the on-going workflow. Once done, check you can access the Web App URL.
 
Build the undeploy example:

```bash
cd examples/undeploy-delete-app
go build -o undeploy
```

undeploy and delete the application:

```bash
./undeploy -url https://1.2.3.4:8088 \
           -user myuser \
           -password mypasswd \
           -app myapp
```

Check undeployment logs appear.
Once done, login on Alien4Cloud UI and confirm the application has been deleted.

### CODE REVIEW

File: alien4cloud/alien4cloud.go [hunk 1]
```diff
@@ -0,0 +1,2169 @@
...
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/goware/urlx"
+	"github.com/pkg/errors"
+)
+
+// Client is the client interface to Alien4cloud service
+type Client interface {
```
Comment (loicalbertin) [inline alien4cloud/alien4cloud.go:37]: > [The bigger the interface, the weaker the abstraction.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=5m17s)

File: examples/undeploy-delete-app/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# Create an application from a template and deploy it on a location
```
Comment (loicalbertin) [inline examples/undeploy-delete-app/README.md]: Seems to be a copy/paste

File: examples/run-workflow/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
+# Run a workflow
+
+This example shows how the Alien4Cloud go client can be used to undeploy
```
Comment (loicalbertin) [inline examples/run-workflow/README.md]: Copy/paste

File: alien4cloud/utils.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+	return errors.New(res.Error.Message)
+}
+
+// ------------------------------------------
+// Implementation of http.CookieJar interface
+// ------------------------------------------
+
+// Jar structure used tO implement http.CookieJar interface
+type Jar struct {
```
Comment (loicalbertin) [inline alien4cloud/utils.go]: Is there a valid reason for Jar type & methods to be public/exported?

File: alien4cloud/alien4cloud_structs.go [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+
+// searchRequest is the representation of a request to search objects as tpologies, orchestrators in the A4C catalog
+type searchRequest struct {
+	Query string `json:"query"`
+	From  string `json:"from"`
+	Size  string `json:"size"`
+}
+
+// environmentsSearchRequest is the representation of a request to search environments of an application in the A4C catalog
+// TODO (HJo): Misnamed type
```
Comment (loicalbertin) [inline alien4cloud/alien4cloud_structs.go]: If misnamed maybe it' time to rename it. Even if I don't know how it should be renamed :smiling_imp:

File: alien4cloud/alien4cloud.go [hunk 6]
```diff
@@ -0,0 +1,2169 @@
...
+
+////////////////////////////////////////////
+// Methods related to workflow management //
+////////////////////////////////////////////
+
+// RunWorkflow runs a4c workflowName workflow for the given a4cAppID and a4cEnvID
+func (c *a4cClient) RunWorkflow(a4cAppID string, a4cEnvID string, workflowName string) (*WorkflowExecution, error) {
+
+	// The Alien4Cloud endpoint to start a workflow in Alien4Cloud is synchronous and for now, never finishes (Alien4Cloud 2.1.0-SM7).
+	go func() {
```
Comment (loicalbertin) [inline alien4cloud/alien4cloud.go]: This routine seems to not be stopped in case of timeout below.

Comment (loicalbertin) [reply alien4cloud/alien4cloud.go]: Maybe using a context may help here.

File: alien4cloud/alien4cloud.go [hunk 7]
```diff
@@ -0,0 +1,2169 @@
...
+
+	if err != nil {
+		return errors.Wrapf(err, "Unable to edit the topology of application '%s' and environment '%s'", a4cCtx.AppID, a4cCtx.EnvID)
+	}
+
+	return nil
+}
+
+// AddRelationship Add a new relationship in the A4C topology
+// not used any more
```
Comment (loicalbertin) [inline alien4cloud/alien4cloud.go]: Remove either the comment or the function

File: alien4cloud/alien4cloud.go [hunk 8]
```diff
@@ -0,0 +1,2169 @@
...
+
+	if err != nil {
+		return errors.Wrapf(err, "Unable to edit the topology of application '%s' and environment '%s'", a4cCtx.AppID, a4cCtx.EnvID)
+	}
+
+	return nil
+}
+
+// AddNodeInA4CTopology Add a new node in the A4C topology
+// not used any more
```
Comment (loicalbertin) [inline alien4cloud/alien4cloud.go]: Remove either the comment or the function

File: alien4cloud/alien4cloud.go [hunk 9]
```diff
@@ -0,0 +1,2169 @@
...
+
+	return nil
+}
+
+///////////////////////////////////////////////////
+// Methods related to application tag management //
+///////////////////////////////////////////////////
+
+// Tag tag key/value json mapping
+type Tag struct {
```
Comment (loicalbertin) [inline alien4cloud/alien4cloud.go]: Why there and not in struct file bellow

File: alien4cloud/alien4cloud.go [hunk 10]
```diff
@@ -0,0 +1,2169 @@
...
+	return templateID, nil
+}
+
+///////////////////////////////////////////////
+// Methods related to application management //
+///////////////////////////////////////////////
+
+// Application represent fields of an application returned by A4C
+// TODO represent more fields of an application returned by A4C
+type Application struct {
```
Comment (loicalbertin) [inline alien4cloud/alien4cloud.go]: Why there and not in the struct file?

File: examples/undeploy-delete-app/README.md (hunk 2 above)
Comment (laurentganne) [reply examples/undeploy-delete-app/README.md]: yes, fixed it

File: alien4cloud/alien4cloud.go (hunk 7 above)
Comment (laurentganne) [reply alien4cloud/alien4cloud.go]: removed the comment

File: alien4cloud/alien4cloud.go (hunk 9 above)
Comment (laurentganne) [reply alien4cloud/alien4cloud.go]: moved it to structs file

File: alien4cloud/alien4cloud.go (hunk 8 above)
Comment (laurentganne) [reply alien4cloud/alien4cloud.go]: Removed the comment

File: alien4cloud/alien4cloud.go (hunk 10 above)
Comment (laurentganne) [reply alien4cloud/alien4cloud.go]: Moved to structs file

File: alien4cloud/alien4cloud_structs.go (hunk 5 above)
Comment (loicalbertin) [reply alien4cloud/alien4cloud_structs.go]: Or remove comment if it doesn't make sense anymore

File: examples/run-workflow/README.md (hunk 3 above)
Comment (laurentganne) [reply examples/run-workflow/README.md]: fixed

File: alien4cloud/alien4cloud_structs.go (hunk 5 above)
Comment (laurentganne) [reply alien4cloud/alien4cloud_structs.go]: Removed this type and its usage, as looking at the REST API, it is exactly the type searchRequest defined above that must be used.

File: alien4cloud/utils.go (hunk 4 above)
Comment (laurentganne) [reply alien4cloud/utils.go]: No need to export type Jar and func NewJar(), fixed it.
Keeping func SetCookies() and Cookies() exported as this type need to implement interface http.CookieJar

File: alien4cloud/alien4cloud.go (hunk 6 above)
Comment (laurentganne) [reply alien4cloud/alien4cloud.go]: Fixed creating a cancelable context and calling http.NewRequestWithContext()

File: alien4cloud/alien4cloud.go (hunk 1 above)
Comment (laurentganne) [reply alien4cloud/alien4cloud.go:37]: Fixed similarly to the kubernetes go client implementation:
the interface exposes now services, application service, deployment service, orchestrator service, topology service, log service

Comment (loicalbertin) [reply alien4cloud/alien4cloud.go:37]: That's such a great work! :heart_eyes: 
Congrats!

**Review by loicalbertin — APPROVED** [review]
LGTM :+1: :clap:

**Pull request merged**